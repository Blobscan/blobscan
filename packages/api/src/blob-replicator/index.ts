import { FlowProducer, Worker } from "bullmq";
import type { FlowChildJob, FlowJob, WorkerOptions } from "bullmq";
import path from "path";

import { createOrLoadBlobStorageManager } from "@blobscan/blob-storage-manager";
import { BlobStorage } from "@blobscan/db";

import { FLOW_PRODUCER_QUEUE, STORAGE_QUEUES } from "./config";
import type { BlobReplicationJobData } from "./types";

const WORKERS_DIR = "workers";

const STORAGE_WORKER_FILES: Record<BlobStorage, string> = {
  GOOGLE: path.join(__dirname, WORKERS_DIR, "gcs.js"),
  SWARM: path.join(__dirname, WORKERS_DIR, "swarm.js"),
  POSTGRES: path.join(__dirname, WORKERS_DIR, "postgres.js"),
};
const PROPAGATOR_WORKER_FILE = path.join(
  __dirname,
  WORKERS_DIR,
  "propagator-flow-producer.js"
);

const replicatorFlowProducer = new FlowProducer();

const BLOB_STORAGES = Object.values(BlobStorage);

export async function setUpBlobReplicationWorkers() {
  const blobStorageManager = await createOrLoadBlobStorageManager();
  const workerOptions: WorkerOptions = {
    useWorkerThreads: true,
  };

  const storageWorkers = BLOB_STORAGES.filter(
    blobStorageManager.hasStorage
  ).reduce<Record<BlobStorage, Worker>>((workers, storageName) => {
    const storageWorker = new Worker(
      STORAGE_QUEUES[storageName],
      STORAGE_WORKER_FILES[storageName],
      workerOptions
    );

    storageWorker.on("completed", (job) => {
      console.log(`Completed job ${job.id} in ${STORAGE_QUEUES[storageName]}`);
    });

    storageWorker.on("failed", (job, err) => {
      console.error(
        `Failed job ${job?.id} in ${STORAGE_QUEUES[storageName]}: ${err}`
      );
    });

    workers[storageName] = storageWorker;

    return workers;
  }, {} as Record<BlobStorage, Worker>);

  if (Object.values(storageWorkers).length) {
    const replicatorWorker = new Worker(
      FLOW_PRODUCER_QUEUE,
      PROPAGATOR_WORKER_FILE,
      workerOptions
    );
  }
}

async function createBlobReplicationJob(
  data: BlobReplicationJobData
): Promise<FlowJob> {
  const blobStorageManager = await createOrLoadBlobStorageManager();
  const versionedHash = data.versionedHash;

  const storageJobs: FlowChildJob[] = BLOB_STORAGES.filter(
    blobStorageManager.hasStorage
  ).map<FlowChildJob>((storage) => {
    const jobId = `${storage}-${versionedHash}`;

    return {
      name: jobId,
      queueName: STORAGE_QUEUES[storage],
      data,
      opts: {
        jobId,
      },
    };
  });

  const jobId = `${FLOW_PRODUCER_QUEUE}-${versionedHash}`;

  return {
    name: jobId,
    queueName: FLOW_PRODUCER_QUEUE,
    data,
    opts: {
      jobId,
    },
    children: storageJobs,
  };
}

export async function queueBlobForReplication(data: BlobReplicationJobData) {
  const blobFlowJob = await createBlobReplicationJob(data);

  replicatorFlowProducer.add(blobFlowJob);
}

export async function queueBlobsForReplication(
  blobReplicationData: BlobReplicationJobData[]
) {
  const blobReplicationFlowJobs = await Promise.all(
    blobReplicationData.map((data) => createBlobReplicationJob(data))
  );

  return replicatorFlowProducer.addBulk(blobReplicationFlowJobs);
}
