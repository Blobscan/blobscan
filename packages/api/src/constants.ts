export const DEFAULT_LIMIT = 50;
