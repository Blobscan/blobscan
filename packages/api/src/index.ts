import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
