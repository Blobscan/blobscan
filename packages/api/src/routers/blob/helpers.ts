import type {
  BlobscanPrismaClient,
  EthUsdPrice,
  ExtendedBlobDataStorageReferenceSelect,
  Prisma,
} from "@blobscan/db";
import { EthUsdPriceModel } from "@blobscan/db/prisma/zod";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import {
  normalizePrismaBlobDataStorageReferencesFields,
  normalizePrismaTransactionFields,
} from "../../utils";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
  prismaBlobOnTransactionSchema,
} from "../../zod-schemas";

const blobDataStorageReferenceSelect = {
  blobStorage: true,
  url: true,
} satisfies ExtendedBlobDataStorageReferenceSelect;

export const extendedBlobDataStorageReferenceArgs: Prisma.Args<
  BlobscanPrismaClient["blobDataStorageReference"],
  "findFirstOrThrow"
> = {
  select: blobDataStorageReferenceSelect,
  orderBy: {
    blobStorage: "asc",
  },
};

export const baseBlobSelect = {
  commitment: true,
  proof: true,
  size: true,
  usageSize: true,
  versionedHash: true,
} satisfies Prisma.BlobSelect;

export const baseBlobOnTransactionSelect = {
  blobHash: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  txHash: true,
  txIndex: true,
} satisfies Prisma.BlobsOnTransactionsSelect;

export type ExtendedBlobDataStorageReference = NonNullable<
  Prisma.Result<
    BlobscanPrismaClient["blobDataStorageReference"],
    { select: typeof blobDataStorageReferenceSelect },
    "findFirst"
  >
>;
type PrismaBlob = Prisma.BlobGetPayload<{
  select: typeof baseBlobSelect;
}> & {
  dataStorageReferences: ExtendedBlobDataStorageReference[];
};

type PrismaBlobOnTransaction = Prisma.BlobsOnTransactionsGetPayload<{
  select: typeof baseBlobOnTransactionSelect;
}>;

export type CompletedPrismaBlob = Prettify<
  PrismaBlob & {
    transactions: Prettify<
      PrismaBlobOnTransaction & {
        block?: ExpandedBlock;
        transaction?: ExpandedTransaction & {
          block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
        };
      }
    >[];
  }
>;

export type CompletedPrismaBlobOnTransaction = Prettify<
  PrismaBlobOnTransaction & {
    blob: {
      dataStorageReferences: ExtendedBlobDataStorageReference[];
    } & ExpandedBlob;
    block?: ExpandedBlock;
    transaction?: ExpandedTransaction & {
      block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
    };
  }
>;
export function createBlobSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobSelect,
    dataStorageReferences: extendedBlobDataStorageReferenceArgs,
    transactions: {
      select: {
        ...baseBlobOnTransactionSelect,
        ...blockExpand,
        ...txExpand,
      },
      orderBy: {
        blockNumber: "desc",
      },
    },
  } satisfies Prisma.BlobSelect;
}

export function createBlobsOnTransactionsSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobOnTransactionSelect,
    blob: {
      select: {
        ...baseBlobSelect,
        dataStorageReferences: extendedBlobDataStorageReferenceArgs,
      },
    },
    ...txExpand,
    ...blockExpand,
  } satisfies Prisma.BlobsOnTransactionsSelect;
}

const responseBlobOnTxSchema = prismaBlobOnTransactionSchema
  .omit({ blobHash: true })
  .required({
    blockHash: true,
    blockNumber: true,
    blockTimestamp: true,
    index: true,
    txHash: true,
    txIndex: true,
  });

export const responseBlobSchema = baseBlobSchema.extend({
  transactions: z.array(
    responseBlobOnTxSchema
      .merge(
        baseTransactionSchema
          .omit({
            hash: true,
            blockNumber: true,
            blockTimestamp: true,
            blockHash: true,
          })
          .partial()
      )
      .extend({
        block: baseBlockSchema
          .omit({
            hash: true,
            number: true,
            timestamp: true,
          })
          .partial({
            blobGasBaseUsdFee: true,
            blobGasUsdPrice: true,
          })
          .optional(),
        ethUsdPrice: EthUsdPriceModel.shape.price.optional(),
      })
  ),
});

export const responseBlobOnTransactionSchema = responseBlobOnTxSchema
  .merge(baseBlobSchema)
  .extend({
    block: baseBlockSchema
      .omit({
        hash: true,
        number: true,
        timestamp: true,
      })
      .optional(),
    transaction: baseTransactionSchema
      .omit({
        hash: true,
        blockNumber: true,
        blockTimestamp: true,
        blockHash: true,
      })
      .optional(),
  });

export type ResponseBlob = z.input<typeof responseBlobSchema>;
export type ResponseBlobOnTransaction = z.input<
  typeof responseBlobOnTransactionSchema
>;

export function toResponseBlob(
  {
    dataStorageReferences,
    transactions: blobOnTxs,
    ...prismaBlob
  }: CompletedPrismaBlob,
  ethUsdPrices: EthUsdPrice["price"][]
): ResponseBlob {
  const responseBlob: ResponseBlob = {
    ...prismaBlob,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    transactions: blobOnTxs.map(
      (
        {
          block: { computeUsdFields, ...restBlock } = {},
          transaction: prismaTx = {},
          ...restBlobOnTx
        },
        i
      ) => {
        const txEthUsdPrice = ethUsdPrices[i];

        const blockUsdFields =
          txEthUsdPrice && computeUsdFields
            ? computeUsdFields(txEthUsdPrice)
            : undefined;
        const txBlock = {
          ...restBlock,
          ...(blockUsdFields ?? {}),
        };

        let responseBlobOnTx: ResponseBlob["transactions"][number] = {
          ...restBlobOnTx,
          ...txBlock,
          ethUsdPrice: txEthUsdPrice?.toNumber(),
        };

        if (isFullyDefined(prismaTx)) {
          const {
            block,
            decodedFields,
            from,
            fromId,
            toId,
            computeBlobGasBaseFee,
            computeUsdFields,
            ...restPrismaTx
          } = prismaTx;

          responseBlobOnTx = {
            ...responseBlobOnTx,
            ...restPrismaTx,
            ...normalizePrismaTransactionFields({
              decodedFields,
              from,
              fromId,
              toId,
            }),
            blobGasBaseFee: computeBlobGasBaseFee(block.blobGasPrice),
            ...(txEthUsdPrice
              ? computeUsdFields({
                  blobGasPrice: block.blobGasPrice,
                  ethUsdPrice: txEthUsdPrice,
                })
              : {}),
          };
        }

        return responseBlobOnTx;
      }
    ),
  };

  return responseBlob;
}

export function toResponseBlobOnTransaction(
  prismaBlobOnTransaction: CompletedPrismaBlobOnTransaction,
  ethUsdPrice?: EthUsdPrice["price"]
): ResponseBlobOnTransaction {
  const {
    blobHash,
    blob: { dataStorageReferences, ...blob },
    transaction,
    block,
    ...restBlobOnTransaction
  } = prismaBlobOnTransaction;

  const responseBlobOnTx: ResponseBlobOnTransaction = {
    ...blob,
    versionedHash: blobHash,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    ...restBlobOnTransaction,
  };

  if (transaction) {
    const {
      block: txBlock,
      decodedFields,
      from,
      fromId,
      toId,
      computeBlobGasBaseFee,
      computeUsdFields,
      ...restTx
    } = transaction;

    responseBlobOnTx.transaction = {
      ...restTx,
      blobGasPrice: txBlock.blobGasPrice,
      ...normalizePrismaTransactionFields({
        decodedFields,
        from,
        fromId,
        toId,
      }),
      blobGasBaseFee: computeBlobGasBaseFee(txBlock.blobGasPrice),
      ...(ethUsdPrice
        ? computeUsdFields({
            blobGasPrice: txBlock.blobGasPrice,
            ethUsdPrice,
          })
        : {}),
    };
  }

  if (block) {
    const { computeUsdFields, ...restBlock } = block;
    responseBlobOnTx.block = {
      ...restBlock,
      ...(ethUsdPrice ? computeUsdFields(ethUsdPrice) : {}),
    };
  }

  return responseBlobOnTx;
}
