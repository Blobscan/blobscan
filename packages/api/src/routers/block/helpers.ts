import type { BlobStorageManager } from "@blobscan/blob-storage-manager";
import type { BlobscanPrismaClient, Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Filters } from "../../middlewares/withFilters";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import { retrieveBlobData } from "../../utils";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
} from "../../zod-schemas";

const prismaBlockSelect = {
  hash: true,
  number: true,
  timestamp: true,
  slot: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  blobGasPrice: true,
  excessBlobGas: true,
} satisfies Prisma.BlockSelect;

export type PrismaBlock = Prisma.BlockGetPayload<{
  select: typeof prismaBlockSelect;
}>;

export type CompletePrismaBlock = Prettify<
  PrismaBlock & {
    transactions: Prettify<
      {
        hash: string;
      } & Partial<ExpandedTransaction> & {
          blobs: Prettify<
            { blobHash: string } & {
              blob?: ExpandedBlob;
            }
          >[];
        }
    >[];
  }
>;

export function createBlockSelect(expands: Expands, filters?: Filters) {
  const blobExpand = expands.blob ? { blob: expands.blob } : {};
  const transactionExpand = expands.transaction?.select ?? {};

  return {
    ...prismaBlockSelect,
    transactions: {
      select: {
        hash: true,
        ...transactionExpand,
        blobs: {
          select: {
            blobHash: true,
            ...blobExpand,
          },
          orderBy: {
            index: "asc",
          },
        },
      },
      orderBy: {
        index: "asc",
      },
      where: filters?.transactionFilters,
    },
  } satisfies Prisma.BlockSelect;
}

export const responseBlockSchema = baseBlockSchema.extend({
  transactions: z.array(
    baseTransactionSchema
      .omit({
        blockHash: true,
        blockNumber: true,
        blockTimestamp: true,
      })
      .partial()
      .required({ hash: true })
      .extend({
        blobs: z.array(
          baseBlobSchema.partial().required({ versionedHash: true })
        ),
      })
  ),
});

export type ResponseBlock = z.input<typeof responseBlockSchema>;

export function toResponseBlock(
  prismaBlock: CompletePrismaBlock
): ResponseBlock {
  const transactions = prismaBlock.transactions.map(
    ({ hash, blobs, ...prismaTx }) => ({
      hash,
      ...(isFullyDefined(prismaTx)
        ? {
            ...prismaTx,
            ...normalizePrismaTransactionFields(prismaTx),
            ...deriveTransactionFields({
              ...prismaTx,
              blobGasPrice: prismaBlock.blobGasPrice,
            }),
          }
        : {}),
      blobs: blobs.map(({ blobHash, blob }) =>
        blob
          ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
          : { versionedHash: blobHash }
      ),
    })
  );

  return {
    ...prismaBlock,
    transactions,
  };
}

export type BlockId = "hash" | "number" | "slot";
export type BlockIdField =
  | { type: "hash"; value: string }
  | { type: "number"; value: number }
  | { type: "slot"; value: number };

function buildBlockWhereClause(
  { type, value }: BlockIdField,
  filters: Filters
): Prisma.BlockWhereInput {
  switch (type) {
    case "hash": {
      return { hash: value };
    }
    case "number": {
      return { number: value, transactionForks: filters.blockType };
    }
    case "slot": {
      return { slot: value, transactionForks: filters.blockType };
    }
  }
}

export async function fetchBlock(
  blockId: BlockIdField,
  {
    blobStorageManager,
    prisma,
    filters,
    expands,
  }: {
    blobStorageManager: BlobStorageManager;
    prisma: BlobscanPrismaClient;
    filters: Filters;
    expands: Expands;
  }
): Promise<CompletePrismaBlock | undefined> {
  const select = createBlockSelect(expands);
  const where = buildBlockWhereClause(blockId, filters);

  const prismaBlock = (await prisma.block.findFirst({
    select,
    where,
  })) as unknown as CompletePrismaBlock | null;

  if (!prismaBlock) {
    return;
  }

  if (expands.blobData) {
    await Promise.all(
      prismaBlock.transactions
        .flatMap((tx) => tx.blobs)
        .map(async ({ blobHash, blob }) => {
          const dataStorageReferences = blob?.dataStorageReferences;
          if (dataStorageReferences) {
            const data = await retrieveBlobData(blobStorageManager, {
              dataStorageReferences,
              versionedHash: blobHash,
            });

            blob.data = data;
          }
        })
    );
  }

  return prismaBlock;
}
