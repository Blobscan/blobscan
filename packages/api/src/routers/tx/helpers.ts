import type { Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseTransactionSchema,
  baseBlobSchema,
  baseBlockSchema,
} from "../../zod-schemas";

const transactionSelect = {
  hash: true,
  fromId: true,
  toId: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  gasPrice: true,
  maxFeePerBlobGas: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  decodedFields: true,
  from: {
    select: {
      address: true,
      rollup: true,
    },
  },
} satisfies Prisma.TransactionSelect;

type PrismaTransaction = Prisma.TransactionGetPayload<{
  select: typeof transactionSelect;
}>;

export type CompletePrismaTransaction = Prettify<
  PrismaTransaction & {
    decodedFields: NonNullable<PrismaTransaction["decodedFields"]>;
    block: Prettify<
      Partial<ExpandedBlock> & {
        blobGasPrice: ExpandedBlock["blobGasPrice"];
      }
    >;
    blobs: Prettify<
      { blobHash: string } & {
        blob?: ExpandedBlob;
      }
    >[];
  }
>;
export function createTransactionSelect(expands: Expands) {
  const blockExpand = expands.block?.select;
  const blobExpand = expands.blob;

  return {
    ...transactionSelect,
    block: {
      select: {
        ...(blockExpand ?? {}),
        blobGasPrice: true,
      },
    },
    blobs: {
      select: {
        blobHash: true,
        ...(blobExpand ? { blob: blobExpand } : {}),
      },
      orderBy: {
        index: "asc",
      },
    },
  } satisfies Prisma.TransactionSelect;
}

export const responseTransactionSchema = baseTransactionSchema.extend({
  block: baseBlockSchema
    .omit({
      hash: true,
      number: true,
      timestamp: true,
    })
    .partial()
    .required({
      blobGasPrice: true,
    }),
  blobs: z.array(baseBlobSchema.partial().required({ versionedHash: true })),
});

export type ResponseTransaction = z.input<typeof responseTransactionSchema>;

export function toResponseTransaction(
  prismaTx: CompletePrismaTransaction
): ResponseTransaction {
  const { blobs: blobsOnTxs, block } = prismaTx;
  const normalizedFields = normalizePrismaTransactionFields(prismaTx);
  const derivedFields = deriveTransactionFields({
    ...prismaTx,
    blobGasPrice: block.blobGasPrice,
  });
  const blobs = blobsOnTxs.map(({ blobHash, blob }) =>
    blob
      ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
      : { versionedHash: blobHash }
  );

  return {
    ...prismaTx,
    ...normalizedFields,
    ...derivedFields,
    blobs,
  };
}
