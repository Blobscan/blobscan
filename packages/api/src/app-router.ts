import { z } from "@blobscan/zod";

import { publicProcedure } from "./procedures";
import { blobRouter } from "./routers/blob";
import { blobStoragesStateRouter } from "./routers/blob-storages-state";
import { blockRouter } from "./routers/block";
import { blockchainSyncStateRouter } from "./routers/blockchain-sync-state";
import { indexerRouter } from "./routers/indexer";
import { searchRouter } from "./routers/search";
import { statsRouter } from "./routers/stats";
import { transactionRouter } from "./routers/tx";
import { t } from "./trpc-client";

export const appRouter = t.router({
  block: blockRouter,
  tx: transactionRouter,
  blob: blobRouter,
  search: searchRouter,
  stats: statsRouter,
  indexer: indexerRouter,
  syncState: blockchainSyncStateRouter,
  blobStoragesState: blobStoragesStateRouter,
  healthcheck: publicProcedure
    .meta({
      openapi: {
        method: "GET",
        path: "/healthcheck",
        summary: "connection healthcheck.",
        tags: ["system"],
      },
    })
    .input(z.void())
    .output(z.string())
    .query(() => "yay!"),
});

// export type definition of API
export type AppRouter = typeof appRouter;
