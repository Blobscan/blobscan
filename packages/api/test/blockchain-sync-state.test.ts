import { beforeAll, describe, expect, it } from "vitest";

import { appRouter } from "../src/app-router";
import { createTestContext, unauthorizedRPCCallTest } from "./helpers";

describe("Blockchain sync state route", async () => {
  let nonAuthorizedCaller: ReturnType<typeof appRouter.createCaller>;
  let authorizedCaller: ReturnType<typeof appRouter.createCaller>;
  let authorizedContext: Awaited<ReturnType<typeof createTestContext>>;

  beforeAll(async () => {
    const ctx = await createTestContext();

    authorizedContext = await createTestContext({ withAuth: true });

    nonAuthorizedCaller = appRouter.createCaller(ctx);
    authorizedCaller = appRouter.createCaller(authorizedContext);
  });

  describe("getState", () => {
    it("should get blockchain sync state", async () => {
      const result = await nonAuthorizedCaller.syncState.getState();
      const expectedState: Awaited<
        ReturnType<typeof nonAuthorizedCaller.syncState.getState>
      > = {
        lastAggregatedBlock: 1004,
        lastFinalizedBlock: 1007,
        lastLowerSyncedSlot: 106,
        lastUpperSyncedSlot: 107,
      };
      expect(result).toMatchObject(expectedState);
    });
  });

  describe("updateState", () => {
    describe("when authorized", () => {
      describe("when updating blockchain sync state", () => {
        it("should update the last finalized block correctly", async () => {
          const prevBlockchainSyncState =
            await authorizedContext.prisma.blockchainSyncState.findUnique({
              where: {
                id: 1,
              },
            });
          const newLastFinalizedBlock = 2000;

          await authorizedCaller.syncState.updateState({
            lastFinalizedBlock: newLastFinalizedBlock,
          });

          const afterBlockchainSyncState =
            await authorizedContext.prisma.blockchainSyncState.findUnique({
              where: {
                id: 1,
              },
            });

          expect(afterBlockchainSyncState).toMatchObject({
            ...prevBlockchainSyncState,
            lastFinalizedBlock: newLastFinalizedBlock,
          });
        });

        it("should update the last lower synced slot correctly", async () => {
          const prevBlockchainSyncState =
            await authorizedContext.prisma.blockchainSyncState.findUnique({
              where: {
                id: 1,
              },
            });
          const newLastLowerSyncedSlot = 2000;

          await authorizedCaller.syncState.updateState({
            lastLowerSyncedSlot: newLastLowerSyncedSlot,
          });

          const afterBlockchainSyncState =
            await authorizedContext.prisma.blockchainSyncState.findUnique({
              where: {
                id: 1,
              },
            });

          expect(afterBlockchainSyncState).toMatchObject({
            ...prevBlockchainSyncState,
            lastLowerSyncedSlot: newLastLowerSyncedSlot,
          });
        });

        it("should update the last upper synced slot correctly", async () => {
          const prevBlockchainSyncState =
            await authorizedContext.prisma.blockchainSyncState.findUnique({
              where: {
                id: 1,
              },
            });
          const newLastUpperSyncedSlot = 2000;

          await authorizedCaller.syncState.updateState({
            lastUpperSyncedSlot: newLastUpperSyncedSlot,
          });

          const afterBlockchainSyncState =
            await authorizedContext.prisma.blockchainSyncState.findUnique({
              where: {
                id: 1,
              },
            });

          expect(afterBlockchainSyncState).toMatchObject({
            ...prevBlockchainSyncState,
            lastUpperSyncedSlot: newLastUpperSyncedSlot,
          });
        });

        it("should update all fields correctly", async () => {
          const newBlockchainSyncState = {
            lastFinalizedBlock: 2001,
            lastLowerSyncedSlot: 30,
            lastUpperSyncedSlot: 560,
          };

          await authorizedCaller.syncState.updateState(newBlockchainSyncState);

          const afterBlockchainSyncState =
            await authorizedContext.prisma.blockchainSyncState.findUnique({
              where: {
                id: 1,
              },
            });

          expect(afterBlockchainSyncState).toMatchObject(
            newBlockchainSyncState
          );
        });
      });

      it("should fail when trying to update last lower synced slot to a value greater than last upper synced slot", async () => {
        const newState = {
          lastLowerSyncedSlot: 1001,
          lastUpperSyncedSlot: 1000,
        };

        await expect(
          authorizedCaller.syncState.updateState(newState)
        ).rejects.toThrowErrorMatchingInlineSnapshot(
          '"Last lower synced slot must be less than or equal to last upper synced slot"'
        );
      });
    });

    unauthorizedRPCCallTest(() =>
      nonAuthorizedCaller.syncState.updateState({
        lastLowerSyncedSlot: 1001,
      })
    );
  });
});
