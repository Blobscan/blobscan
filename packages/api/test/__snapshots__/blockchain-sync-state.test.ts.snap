// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last finalized slot to an invalid slot 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 0,
    \\"type\\": \\"number\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Number must be greater than or equal to 0\\",
    \\"path\\": [
      \\"lastFinalizedBlock\\"
    ]
  }
]"
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last finalized slot to an invalid slot 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be greater than or equal to 0",
    "path": [
      "lastFinalizedBlock"
    ]
  }
]]
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last lower synced slot to an invalid slot 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 0,
    \\"type\\": \\"number\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Number must be greater than or equal to 0\\",
    \\"path\\": [
      \\"lastLowerSyncedSlot\\"
    ]
  }
]"
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last lower synced slot to an invalid slot 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be greater than or equal to 0",
    "path": [
      "lastLowerSyncedSlot"
    ]
  }
]]
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last upper synced block root to an invalid hash 1`] = `
"[
  {
    \\"validation\\": \\"regex\\",
    \\"code\\": \\"invalid_string\\",
    \\"message\\": \\"Invalid hexadecimal string\\",
    \\"path\\": [
      \\"lastUpperSyncedBlockRoot\\"
    ]
  },
  {
    \\"code\\": \\"custom\\",
    \\"message\\": \\"Block hashes must be 66 characters long\\",
    \\"path\\": [
      \\"lastUpperSyncedBlockRoot\\"
    ]
  }
]"
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last upper synced block root to an invalid hash 2`] = `
[ZodError: [
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid hexadecimal string",
    "path": [
      "lastUpperSyncedBlockRoot"
    ]
  },
  {
    "code": "custom",
    "message": "Block hashes must be 66 characters long",
    "path": [
      "lastUpperSyncedBlockRoot"
    ]
  }
]]
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last upper synced block slot to an invalid slot 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 0,
    \\"type\\": \\"number\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Number must be greater than or equal to 0\\",
    \\"path\\": [
      \\"lastUpperSyncedBlockSlot\\"
    ]
  }
]"
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last upper synced block slot to an invalid slot 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be greater than or equal to 0",
    "path": [
      "lastUpperSyncedBlockSlot"
    ]
  }
]]
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last upper synced slot to an invalid slot 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 0,
    \\"type\\": \\"number\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Number must be greater than or equal to 0\\",
    \\"path\\": [
      \\"lastUpperSyncedSlot\\"
    ]
  }
]"
`;

exports[`Blockchain sync state route > updateState > when authorized > should fail when trying to update last upper synced slot to an invalid slot 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be greater than or equal to 0",
    "path": [
      "lastUpperSyncedSlot"
    ]
  }
]]
`;
